### ISSUE TITLE
Should the stdio buffer size be increased for the dump code?

### ISSUE THREAD

**Issue by guyharris** [the-tcpdump-group/libpcap#792]
To report a security issue please send an e-mail to security@tcpdump.org.

To report bugs and other problems, contribute patches, request a
feature, provide generic feedback etc please see the file
CONTRIBUTING in the libpcap source tree root.

**Comment by guyharris** [comment]
He did use a larger buffer - 16KiB.  That was probably on Linux; the default `st_blksize` on Linux appears to be the page size, so it was probably 4KiB.  If a 16KiB `fwrite()` bypasses the buffer and just writes directly to the file, as a result of being a multiple of the buffer size, then that'd be a 4x reduction in the number of `write()` calls and thus a 4x reduction in the number of system calls, which might make a significant difference.

A bigger buffer would result in an even greater reduction, although I don't know whether a buffer that's close to or bigger than the L1 data cache (32K on most of Intel's x86-64 processors) would be a problem.

**Comment by guyharris** [comment]
As for a manually-allocated buffer:

- [The Linux `setvbuf()` documentation](http://man7.org/linux/man-pages/man3/setvbuf.3.html), which presumably is describing the glibc version, "If the argument *`buf`* is `NULL`, only the mode is        affected; a new buffer will be allocated on the next read or write operation.".  [The GNU libc documentation for `setvbuf()`](https://www.gnu.org/software/libc/manual/html_mono/libc.html#Controlling-Buffering), however, just says "If you specify a null pointer as the *buf* argument, then `setvbuf` allocates a buffer itself using `malloc`. This buffer will be freed when you close the stream. ... Otherwise, *buf* should be a character array that can hold at least *size* characters.", which doesn't explicitly indicate that the buffer allocated by `setvbuf()` will have size *size*.
- [The MUSL libc documenation](https://www.musl-libc.org/doc/1.0.0/manual.html) just says that the official documentation is the C standard and the POSIX standard and that "This portion of the manual is incomplete. Future editions will document musl's behavior where the standards specify that it is implementation-defined, non-standard extensions musl implements, and additional properties of musl's implementation.", so they don't say one way or the other whether `setvbuf()` has been extended to support setting the buffer size without explicitly setting the buffer.
- [The Bionic libc documentation](https://android.googlesource.com/platform/bionic/+/HEAD/docs) says nothing, but [the Bionic `setvbuf()` source](https://android.googlesource.com/platform/bionic/+/fe6338d/libc/stdio/setvbuf.c) indicates that it - and probably the rest of the standard I/O library - is BSD-derived, so it probably implements the same "NULL argument plus non-zero size plus not unbuffered means a buffer of the specified size will be automatically allocated" behavior, and that's what the code appears to do.
- [The FreeBSD 3.0 `setvbuf()` documentation](https://www.freebsd.org/cgi/man.cgi?query=setvbuf&apropos=0&sektion=0&manpath=FreeBSD+3.0-RELEASE&arch=default&format=html) says "If the *`size`* argument is not zero but *`buf`* is `NULL`, a buffer of the given size will be allocated immediately, and released on close.".
- [The NetBSD 2.0](http://netbsd.gw.com/cgi-bin/man-cgi?setvbuf++NetBSD-2.0), [OpenBSD 2.2](https://man.openbsd.org/OpenBSD-2.2/setvbuf), [DragonFly BSD](https://leaf.dragonflybsd.org/cgi/web-man?command=setvbuf&section=ANY), and [macOS](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/setvbuf.3.html) `setvbuf()` documentation say the same thing (not surprising, given that they probably all started with the 4.4-Lite libc, and that behavior probably dates back a long time for all those OSes).
- [The Visual Studio 2015 `setvbuf()` documentation](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/setvbuf?view=vs-2015) says "The array pointed to by *buffer* is used as the buffer, unless it is NULL, in which case **setvbuf** uses an automatically allocated buffer of length *size*/2 \* 2 bytes.".

However:

- [the Solaris 11.4 `setvbuf()` man page](https://docs.oracle.com/cd/E88353_01/html/E37843/setbuf-3c.html#scrolltoc) just says "If `buf` is not the null pointer, the array it points to will be used for buffering, instead of an automatically allocated buffer. The `size` argument specifies the size of the buffer to be used. If input/output is unbuffered, `buf` and `size` are ignored.", which could be read as suggesting that if `buf` *is* the null pointer, and the stream is buffered, it automatically allocates a buffer of `size` bytes, but they don't unequivocally state that.
- [the AIX 5.3 `setvbuf()` man page](http://public.dhe.ibm.com/systems/power/docs/aix/53/basetrf2.pdf) says "If the *Buffer* parameter is not a null character pointer, the array it points to is used for buffering. The *Size* parameter specifies the size of the array to be used as a buffer, but all of the *Size* parameter’s bytes are not necessarily used for the buffer area.", which could again be read as suggesting that a null *Buffer* argument, non-zero *Size* argument, and *Mode* argument other than `_IONBF` means that a buffer of the specified size will automatically be allocated, but they don't say it explicitly;
- [the HP-UX 11i v3 `setvbuf()` man page](https://support.hpe.com/hpsc/doc/public/display?sp4ts.oid=4342057&docLocale=en_US&docId=emr_na-c01922517&withFrame) says "If *buf* is not the NULL pointer, the array it points to is used for buffering instead of an automatically allocated buffer (from malloc()). size specifies the size of the buffer to be used. *size* specifies the size of the buffer to be used.", which again suggests that you can set the size of an automatically-allocated buffer but doesn't explicitly state it;
- [the Tru64 UNIX 5.1 `setvbuf()` man page](http://web.archive.org/web/20040702192619/http://h30097.www3.hp.com:80/docs/base_doc/DOCUMENTATION/V51B_HTML/MAN/MAN3/0536____.HTM) says "If the *`buffer`* parameter is not a null pointer, the array that the parameter points to is used for buffering instead of a buffer that is automatically allocated. The *`size`* parameter specifies the size of the buffer to be used." - you know the drill;
- [the IRIX 6.5 `setvbuf()` man page](http://web.archive.org/web/20160119195454/http://techpubs.sgi.com/library/tpl/cgi-bin/getdoc.cgi?coll=0650&db=man&fname=/usr/share/catman/p_man/cat3s/setbuf.z) says "If *buf* is the NULL pointer, or *size* is less than eight, space will be allocated to accommodate a buffer.  This buffer will be of length BUFSIZ.", which means that 1) the good news is that a NULL *buf* means it'll automatically allocate a buffer but 2) the bad news is that, no matter what the value of *size* is, the buffer will be BUFSIZ bytes.

**Comment by AndersBroman** [comment]
Something like this?
```diff
--- a/sf-pcap.c
+++ b/sf-pcap.c
@@ -771,7 +771,11 @@ pcap_setup_dump(pcap_t *p, int linktype, FILE *f, const char *fname)
        else
                setvbuf(f, NULL, _IONBF, 0);
 #endif
-       if (sf_write_header(p, f, linktype, p->snapshot) == -1) {
+#define IO_BUF_SIZE (16 * 1024)
+    if (f != stdout) {
+        setvbuf(f, NULL, _IOFBF, IO_BUF_SIZE);
+    }
+        if (sf_write_header(p, f, linktype, p->snapshot) == -1) {
                pcap_fmt_errmsg_for_errno(p->errbuf, PCAP_ERRBUF_SIZE,
                    errno, "Can't write to %s", fname);
                if (f != stdout)
```

**Comment by guyharris** [comment]
Yeah, the Linux man page isn't as explicit as I'd like, but "If the argument `*buf*` is `NULL`, only the mode is affected" does suggest that the size isn't affected.

If an explicit buffer is given, we don't have any good way to free it in `pcap_dump_close()`, given that:

1. a `pcap_dumper_t *` is just another name for a `FILE *`, so we don't have any place to save a pointer to the buffer;
1. there's code out there that needs to, for example, do an `fflush()` on the file being written, and does so by doing the `fflush()` on the `pcap_dumper_t *` cast to a `FILE *` if `pcap_dump_flush()` isn't available, so it's risky to change what a `pcap_dumper_t` is;
1. there's no generally-available API to get a pointer to the buffer.

So the choices we have are:

1. don't do this on Linux;
1. develop new APIs for dumping, with a handle type different from `pcap_dumper_t *` that points to an opaque structure that includes the `FILE *` and a `void *` pointing to the buffer.

I'm inclined to go for the second option.

**Comment by mcr** [comment]
Guy Harris <notifications@github.com> wrote:
    > 1 a pcap_dumper_t * is just another name for a FILE *, so we don't have
    > any place to save a pointer to the buffer; 2 there's code out there
    > that needs to, for example, do an fflush() on the file being written,
    > and does so by doing the fflush() on the pcap_dumper_t * cast to a FILE

I guess this was a bad thing.
Such code would fail to recompile, but of course it's an ABI change.

    > * if pcap_dump_flush() isn't available, so it's risky to change what a
    > pcap_dumper_t is; 3 there's no generally-available API to get a pointer
    > to the buffer.

:-(

    > So the choices we have are:

    > 1 don't do this on Linux;
    > 2 develop new APIs for dumping, with a handle
    > type different from pcap_dumper_t * that points to an opaque structure
    > that includes the FILE * and a void * pointing to the buffer.

    > I'm inclined to go for the second option.

Me too.
We are going to need it for pcapng.

--
]               Never tell me the odds!                 | ipv6 mesh networks [
]   Michael Richardson, Sandelman Software Works        |    IoT architect   [
]     mcr@sandelman.ca  http://www.sandelman.ca/        |   ruby on rails    [

**Comment by guyharris** [comment]
> We are going to need it for pcapng.

Yes.  I wouldn't want to require pcapng to get the bigger stdio buffer, and I'm not sure whether we should have a single new API for writing both pcap and pcapng files, but either way we should have support for writing pcap files, not just pcapng files, with a bigger stdio buffer.

**Comment by Lekensteyn** [comment]
`pcap_dumper_t` is already an opaque structure (already since its initial introduction in 1999), the fact that it is type-casted from `FILE *` is an internal implementation detail. Are there any issues with converting this to an actual structure?

**Comment by guyharris** [comment]
> `pcap_dumper_t` is already an opaque structure (already since its initial introduction in 1999), the fact that it is type-casted from `FILE *` is an internal implementation detail. Are there any issues with converting this to an actual structure?

To quote an earlier comment:

> there's code out there that needs to, for example, do an `fflush()` on the file being written, and does so by doing the `fflush()` on the `pcap_dumper_t *` cast to a `FILE *` if `pcap_dump_flush()` isn't available, so it's risky to change what a `pcap_dumper_t` is;

**Comment by mcr** [comment]
I'm unclear if there is still work to do here.

**Comment by Oppen** [comment]
I think not, and the lack of activity during 5 months backs me up.
Can it be closed?

**Comment by mcr** [comment]
We are going to do the setvbuf() in case it helps.  Ticket remains open for now.
